crates/topos-test-sdk/src/integration/mod.rs

```diff
@@ -0,0 +1,154 @@
+                    .await
+                    .map_err(|_| (peer_string.clone(), "Unable to connect to the api console"))?;
+
+                let result = client.status(StatusRequest {}).await.map_err(|_| {
+                    (
+                        peer_string.clone(),
+                        "Unable to get the status from the api console",
+                    )
+                })?;
+
+                let status = result.into_inner();
+                if !status.has_active_sample {
+                    return Err((peer_string, "didn't succeed in the sample phase"));
+                }
+
+                let mut client = ApiServiceClient::connect(peer_string.clone())
+                    .await
+                    .map_err(|_| (peer_string.clone(), "Unable to connect to the TCE api"))?;
+
+                let in_stream = async_stream::stream! {
+                    yield OpenStream {
+                        target_checkpoint: Some(TargetCheckpoint {
+                            target_subnet_ids: vec![[2u8; SUBNET_ID_LENGTH].into()],
+                            positions: vec![]
+                        }),
+                        source_checkpoint: None
+                    }.into()
+                };
+
+                let response = client.watch_certificates(in_stream).await.map_err(|_| {
+                    (
+                        peer_string.clone(),
+                        "Unable to execute the watch_certificates on TCE api",
+                    )
+                })?;
+                let mut resp_stream = response.into_inner();
+                async move {
+                    while let Some(received) = resp_stream.next().await {
+                        let received = received.unwrap();
+                        if let Some(Event::CertificatePushed(CertificatePushed {
+                            certificate: Some(certificate),
+                            ..
+                        })) = received.event
+                        {
+                            // unwrap is safe because we are sure that the certificate is present
+                            if certificate_id == certificate.id.unwrap() {
+                                debug!("Received the certificate on {}", peer_string);
+                                return Ok(());
+                            }
+                        }
+                    }
+
+                    Err((peer_string.clone(), "didn't receive any certificate"))
+                }
+                .await
+            }));
+        }
+
+        let mut client = ApiServiceClient::connect(random_peer.clone())
+            .await
+            .map_err(|_| vec![format!("Unable to connect to the TCE api on {random_peer}")])?;
+
+        // submit a certificate to one node
+        _ = client
+            .submit_certificate(SubmitCertificateRequest {
+                certificate: Some(pushed_certificate.into()),
```

Comment:
Is this the first use of `pushed_certificate`? Maybe move the creation of that variable closer to where its used?

Reply:
It has to be up because `certificate_id` is used in spawned task